controllers/agents/svc/serviceclass_controller.go

```diff
@@ -95,15 +109,57 @@ func (r *ServiceClassReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
 
 	// then, write all the registered services up to the primaza cluster
-	err = r.CreateRegisteredServices(ctx, &serviceClass, *services)
-	if err != nil {
-		reconcileLog.Error(err, "Failed to write registered services")
-		// fallthrough: we still want to write the service class status field
-	}
+	if serviceClass.DeletionTimestamp.IsZero() {
+		handler := func(remote_client client.Client, rs v1alpha1.RegisteredService) error {
+			op, err := controllerutil.CreateOrUpdate(ctx, remote_client, &rs, func() error { return nil })
+			if err != nil {
+				reconcileLog.Error(err, "Failed to create registered service",
+					"service", rs.Name,
+					"namespace", rs.Namespace)
+			} else {
+				reconcileLog.Info("Wrote registered service", "service", rs.Name, "namespace", rs.Namespace, "operation", op)
+			}
+			return err
+		}
 
-	if err = r.setOwnerReference(ctx, &serviceClass, req.Namespace); err != nil {
-		reconcileLog.Error(err, "Failed to set owner reference on ServiceClass", "namespace", req.Namespace, "name", req.Name)
-		return ctrl.Result{}, err
+		// add a finalizer since we have deletion logic
+		if controllerutil.AddFinalizer(&serviceClass, finalizer) {
+			if err = r.Update(ctx, &serviceClass, &client.UpdateOptions{}); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		err = r.HandleRegisteredServices(ctx, &serviceClass, *services, handler)
+		if err != nil {
+			reconcileLog.Error(err, "Failed to write registered services")
+			// fallthrough: we still want to write the service class status field
+		}
+	} else if controllerutil.ContainsFinalizer(&serviceClass, finalizer) {
+		handler := func(remote_client client.Client, rs v1alpha1.RegisteredService) error {
+			if err := remote_client.Delete(ctx, &rs); err != nil {
+				if apierrors.IsNotFound(err) {
+					// we tried to delete an object that doesn't exist, so
+					return nil
+				}
+				reconcileLog.Error(err, "Failed to delete registered service", "namespace", rs.Namespace)
+			}
+			return err
+		}
+
+		err = r.HandleRegisteredServices(ctx, &serviceClass, *services, handler)
+		if err != nil {
+			reconcileLog.Error(err, "Failed to delete registered services")
+			return ctrl.Result{}, err
+		}
+
+		// remove the finalizer so we don't requeue
+		if controllerutil.RemoveFinalizer(&serviceClass, finalizer) {
+			if err = r.Update(ctx, &serviceClass, &client.UpdateOptions{}); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		// nothing to do
+		return ctrl.Result{}, nil
 	}
 
 	// finally, write the status of the service class
```

Comment:
This part of updating the status of service class is unreachable, so i guess we can remove the else block and proceed with status update or return nil error